warehouse/packaging/tasks.py

```diff
@@ -357,90 +361,69 @@ def sync_bigquery_release_files(request):
     table_names = release_files_table.split()
 
-    for table_name in table_names:
-        table_schema = bq.get_table(table_name).schema
+    missing_files = (
+        request.db.query(MissingDatasetFile)
+        .filter(MissingDatasetFile.processed.is_(False))
+        .limit(10)
+    )
 
-        # Using the schema to populate the data allows us to automatically
-        # set the values to their respective fields rather than assigning
-        # values individually
-        def populate_data_using_schema(file):
-            release = file.release
-            project = release.project
-
-            row_data = dict()
-            for sch in table_schema:
-                # The order of data extraction below is determined based on the
-                # classes that are most recently updated
-                if hasattr(file, sch.name):
-                    field_data = getattr(file, sch.name)
-                elif hasattr(release, sch.name) and sch.name == "description":
-                    field_data = getattr(release, sch.name).raw
-                elif sch.name == "description_content_type":
-                    field_data = getattr(release, "description").content_type
-                elif hasattr(release, sch.name):
-                    field_data = getattr(release, sch.name)
-                elif hasattr(project, sch.name):
-                    field_data = getattr(project, sch.name)
-                else:
-                    field_data = None
-
-                if isinstance(field_data, datetime.datetime):
-                    field_data = field_data.isoformat()
-
-                # Replace all empty objects to None will ensure
-                # proper checks if a field is nullable or not
-                if not isinstance(field_data, bool) and not field_data:
-                    field_data = None
-
-                if field_data is None and sch.mode == "REPEATED":
-                    row_data[sch.name] = []
-                elif field_data and sch.mode == "REPEATED":
-                    # Currently, some of the metadata fields such as
-                    # the 'platform' tag are incorrectly classified as a
-                    # str instead of a list, hence, this workaround to comply
-                    # with PEP 345 and the Core Metadata specifications.
-                    # This extra check can be removed once
-                    # https://github.com/pypi/warehouse/issues/8257 is fixed
-                    if isinstance(field_data, str):
-                        row_data[sch.name] = [field_data]
+    for missing_file in missing_files:
+        release_file = missing_file.file
+
+        for table_name in table_names:
+            table_schema = bq.get_table(table_name).schema
+
+            # Using the schema to populate the data allows us to automatically
+            # set the values to their respective fields rather than assigning
+            # values individually
+            def populate_data_using_schema(file):
```

Comment:
Just a nit, and I'm sure it's annoying, but I am struggling to feel certain that the way we are populating the object here and the way we are populating it during upload are indeed identical.

In my dream world, the [construction of dist_metadata from forklift/legacy/upload](https://github.com/pypi/warehouse/blob/main/warehouse/forklift/legacy.py#L1456) was a function that took a `File` object and returned the dict to pass to the `update_bigquery_release_files` task.

Maybe a bit overkill but otherwise there's just a lot to consider to ensure that we're populating the table in the same way as the mainline task.

Reply:
Unfortunately (as you've discovered) there is data we send to BigQuery which only comes off the parsed metadata and isn't stored in the database (e.g., the metadata version), so there is going to be some slight differences here.